### PR TITLE
fix: treat "0" and "1" as valid boolean values.

### LINKED ISSUE #3368: Failed to parse 0 and 1 as boolean at insertion
### What type of bug is this?

User Experience

### What subsystems are affected?

Standalone mode

### Minimal reproduce step

First create the table in a standalone greptimedb:
```sql
create table foo(ts timestamp time index, b boolean)
```
Use the following java code to insert values:
```java
String sql = "insert into foo(ts, b) values(?,?)";

try (Connection conn = DriverManager.getConnection("jdbc:mysql://127.0.0.1:4002")) {
  PreparedStatement ps = conn.prepareStatement(sql);

  ps.setObject(1, System.currentTimeMillis()); // ts
  ps.setObject(2, true); // b

  ps.addBatch();
  ps.executeBatch();
  ps.clearBatch();
}
```
> With mysql jdbc driver
> ```xml
> <dependency>
>   <groupId>mysql</groupId>
>   <artifactId>mysql-connector-java</artifactId>
>   <version>8.0.33</version>
> </dependency>
> ```
Then the db complains:
```text
2024-02-23T10:02:43.576713Z DEBUG on_query: servers::mysql::writer: Failed to handle mysql query, error: 0: Failed to execute query, query: insert into foo(ts, b) values(1708682506490,1), at /Users/luofucong/repo/greptime/my-greptimedb/src/servers/src/query_handler/sql.rs:89:22
1: Table operation error, at src/frontend/src/instance.rs:270:14
2: Failed to parse SQL, at src/operator/src/req_convert/insert/stmt_to_region.rs:199:14
3: Failed to parse value: Fail to parse number 1, invalid column type: Boolean(BooleanType)
```

### What did you expect to see?

Successfully insert the values, since the client codes are valid.

### What did you see instead?

Failed to insert.

### What operating system did you use?

macos

### What version of GreptimeDB did you use?

main

### Relevant log output and stack trace

_No response_

### PR DESCRIPTION
**Pull request by MichaelScofield**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

fix #3368 

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

File: src/sql/src/statements.rs [hunk 1]
```diff
@@ -172,7 +172,15 @@ macro_rules! parse_number_to_value {
                     }.fail()
                 }
             }
-
+            // It's valid for MySQL JDBC to send "0" and "1" for boolean types, so adapt to that.
+            ConcreteDataType::Boolean(_) => {
+                match $n {
+                    "0" => Ok(Value::Boolean(false)),
```
[the remark this answers is not in the document]
Comment (tisonkun) [reply src/sql/src/statements.rs:178]: This is kind of MySQL's dialet so it's still an unterminated topic if we'd like to be "compatible" in this level. Or just specially handle the "0"/"1" case.

More concretely, MySQL store boolean value in a tinyint column so if you run:

```
CREATE TABLE foo(b boolean);
INSERT INTO foo VALUES (3.5);
SELECT * FROM foo;
```

You will get:

```
+------+
| b    |
+------+
|    4 |
+------+
```

But this doesn't seem what we should be compatible with.

File: src/sql/src/statements.rs [hunk 2]
```diff
@@ -172,7 +172,15 @@ macro_rules! parse_number_to_value {
...
+                    "1" => Ok(Value::Boolean(true)),
+                    _ => ParseSqlValueSnafu {
+                        msg: format!("Failed to parse number '{}' to boolean column type!", $n)}.fail(),
```
Comment (tisonkun) [inline src/sql/src/statements.rs:181]:
```suggestion
                        msg: format!("Failed to parse number '{}' to boolean column type", $n)}.fail(),
```

nit: we may not use this tone else where.

**Review by tisonkun — APPROVED** [review]
I'm OK we're specially handling this MySQL JDBC manner.

Comments inline for records.

File: src/sql/src/statements.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (MichaelScofield) [reply src/sql/src/statements.rs:178]: I don't think we should do this. Mysql does not have built-in boolean data type, so it use tinyint for that, see [here](https://dev.mysql.com/doc/refman/8.0/en/other-vendor-data-types.html). I believe it's the reason why jdbc driver choose to encode java boolean values to 0 and 1. It's done implicitly, so db won't get "0.0" or "1.0".

**Review by WenyXu — APPROVED** [review]

**Pull request merged**